server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java

```diff
@@ -576,6 +577,18 @@ private ClusterState applyCreateIndexRequestWithExistingMetadata(final ClusterSt
                 Map<String, Object> innerTemplateNonProperties = new HashMap<>(innerTemplateMapping);
                 Map<String, Object> maybeProperties = (Map<String, Object>) innerTemplateNonProperties.remove("properties");
 
+                List<Map<String, Object>> innerTemplateDynamicTemplates = (List<Map<String, Object>>) innerTemplateNonProperties.get(
+                    "dynamic_templates");
+                List<Map<String, Object>> previouslySeenDynamicTemplates =
+                    (List<Map<String, Object>>) nonProperties.get("dynamic_templates");
+                // we want "later" defined dynamic templates to override the previously seen ones with the same name
+                List<Map<String, Object>> filteredDefaultDynamicTemplates =
+                    removeAll(previouslySeenDynamicTemplates, innerTemplateDynamicTemplates);
+
+                if (filteredDefaultDynamicTemplates != previouslySeenDynamicTemplates) {
+                    nonProperties.put("dynamic_templates", filteredDefaultDynamicTemplates);
+                }
```

Comment:
I think we can encapsulate this logic into a method to make it a bit clearer what's going on, perhaps something like:

```
nonProperties = removeDuplicatedDynamicTemplates(nonProperties, innerTemplateNonProperties);
```

Then the same method can be used below (where these lines are almost identical except for the maps passed in).